rqueue-spring-boot-starter/src/main/java/com/github/sonus21/rqueue/spring/boot/RqueueListenerAutoConfig.java

```diff
@@ -36,32 +38,40 @@
 import org.springframework.context.annotation.ComponentScan;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.DependsOn;
+import org.springframework.messaging.converter.MessageConverter;
+import java.util.List;
 
 @Configuration
 @AutoConfigureAfter(RedisAutoConfiguration.class)
 @ComponentScan("com.github.sonus21.rqueue.web")
 public class RqueueListenerAutoConfig extends RqueueListenerBaseConfig {
 
+  @Bean
+  @ConditionalOnMissingBean
+  public MessageConverter messageConverter() {
```

Comment:
There's a reason this bean was not created, if the application has already another message converter then it will force us to use and the application user may not be the correct one. Also it will put a message converter bean for any usage, if the application wants to override than it will have to use override flag to override message converter.

Reply:
The Sping Message Object is very simple, here is the whole class:

```
/**
 * A generic message representation with headers and body.
 *
 * @author Mark Fisher
 * @author Arjen Poutsma
 * @since 4.0
 * @param <T> the payload type
 * @see org.springframework.messaging.support.MessageBuilder
 */
public interface Message<T> {

	/**
	 * Return the message payload.
	 */
	T getPayload();

	/**
	 * Return message headers for the message (never {@code null} but may be empty).
	 */
	MessageHeaders getHeaders();

}
```

Outside of tests  Rqueue is receiving a Message<?> only twice and in both cases treating it in similar ways.

Here's 1 example, the other is very similar.
```
public static RqueueMessage buildMessage(
      MessageConverter messageConverter,
      String queueName,
      Object message,
      Integer retryCount,
      Long delayInMilliSecs) {
    Message<?> msg = messageConverter.toMessage(message, null);
    if (msg == null) {
      throw new MessageConversionException("Message could not be build (null)");
    }
    return new RqueueMessage(queueName, (String) msg.getPayload(), retryCount, delayInMilliSecs);
  }

```


In this case Rqueue receives a `Message<?>` object, and throws away the header and everything else and just reads (String) msg.getPayload().

If Rqueue only needs a String then why ask for a whole Message object which can contain any type (not guaranteed to be a String) plus a whole header? If Rqueue only needs a String and doesn't want to override or conflict with other configurations of Spring MessageConverer, then it should define it's own conversion interface:

Something like this would work:

```
public interface RqueueMessageMarshaller {

    String serialize(Object payload);

    Object deserialize(String serialization);
}
```

This would solve the possible conflict and also is a simpler interface to understand.

Reply:
Well, RqueueMessageMarshaller interface can not be used with spring-messaging library. Please check initArgumentResolvers method. Everything boils down to this method, Rqueue does not care about any metadata yet as we don't have to do anything but other message converter can use MessageHeaders to do anything. FYI, Rqueue does not store any headers in Redis, it just build some data on top of the deserialised RqueueMessage.